### PR TITLE
WIP: Add BoxGraph for strongly connected components and transition matrix

### LINKED ISSUE #6: Tie a transition matrix to a BoxSet?
As it stands, a `BoxSet` is unordered, so by assigning an ordering, we introduce additional structure.  We do not necessarily need this when constructing a graph (i.e. for computing strongly connected components), but we **do** need it when computing a transition (probability) matrix.  The problem is - and this has been a problem since the early days - that the information that one typically computes from a transition matrix (e.g. eigenvectors) is later linked back to the `BoxSet` (by, e.g., assiging a color to each `Box`according to the corresponding entry in the eigenvector), so one must not change the `BoxSet` in the meantime or otherwise the information from the transition matrix is no longer valid for the `BoxSet`.  So the question is whether we should somehow 'tie' a transition matrix to a `BoxSet`, i.e. whether we should define something like `BoxTransitionMatrix` which, e.g., takes a `BoxSet` and a `BoxMap` for construction.

_Originally posted by @gaioguy in https://github.com/gaioguy/GAIO.jl/pull/1#issuecomment-654901328_

### PR DESCRIPTION
**Pull request by cafaxo**
Reopen #18 since it was closed by accident. Fixes #6 and #14.

### CODE REVIEW

**Comment by FrederikSchnack** [comment]
For the visualisation of the transition graph of a invariant set, we could consider plotting the boxes with opacity according to their weight in the graph.

**Comment by gaioguy** [comment]
Hm, not sure whether I understand: In the transition graph, the edges are weighted, not the nodes.  And further: What do the weights tell us about the dynamics?

**Comment by FrederikSchnack** [comment]
I was referencing the visualisation in the end of section 4 of [Dellnitz, M.; Junge, O.: Set oriented numerical methods for dynamical systems,
B. Fiedler, G. Iooss and N. Kopell (eds.): Handbook of Dynamical Systems III: Towards Applications, World Scientific, 2002.](http://www-m3.ma.tum.de/twiki/pub/M3/Allgemeines/Publications/handbook.pdf)

**Comment by gaioguy** [comment]
Sure, yes, this is something we should ultimately be aming for!

File: src/boxgraph.jl [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    sizehint!(V, num_edges)
+
+    for (edge, weight) in boxgraph.edges
+        push!(I, key_to_int[edge[1]])
+        push!(J, key_to_int[edge[2]])
+        push!(V, weight)
+    end
+
+    n = length(int_to_key)
+    mat = sparse(I, J, V, n, n)
```
Comment (FrederikSchnack) [inline src/boxgraph.jl:74]: Do we need to use `SparseArrays` in GAIO for that? For me, it wasn't working without it.

File: src/GAIO.jl [hunk 2]
```diff
@@ -12,7 +12,7 @@ export boxset_empty, subdivide, subdivide!
 export BoxPartition, RegularPartition, TreePartition
 export dimension, depth
 
-export BoxMap, PointDiscretizedMap
+export BoxMap, PointDiscretizedMap, transition_graph
```
Comment (FrederikSchnack) [inline src/GAIO.jl:15]: Do we want to export the `matrix` function? This could be convenient.

**Comment by cafaxo** [comment]
Superseded by the idea at https://github.com/gaioguys/GAIO.jl/pull/32#issuecomment-663098767.

**Comment by gaioguy** [comment]
Maybe we should still leave this open.  I think the graph is there to stay, maybe in a slightly different form.

**Comment by cafaxo** [comment]
A new implementation deserves a new PR. We should preserve this old approach here.

**Comment by cafaxo** [comment]
By the way: It is not good practice to push onto another one's branch like this. In Git, everyone works on a particular branch and then asks (using a PR) to merge this branch into master. Merge conflicts are then resolved when the branch is merged with master... If I am locally working on my branch, I wouldn't want to suddenly have to resolve merge conflicts when pushing my branch to the remote.
Here, it would have been better to create your own branch on top of my `graph` branch.
Alternatively, you can use github's review system to place code suggestions in a PR, which I then could apply to my branch.

**Comment by gaioguy** [comment]
Apologies - and thanks for explaining the workflow.  Shall I remove my commit?

**Comment by cafaxo** [comment]
Not necessary since I am no longer working on that branch.

**Pull request closed without merging**